sql/core/src/main/scala/org/apache/spark/sql/execution/SparkOptimizer.scala

```diff
@@ -39,6 +39,9 @@ class SparkOptimizer(
     // TODO: move SchemaPruning into catalyst
     SchemaPruning :: V2ScanRelationPushDown :: PruneFileSourcePartitions :: Nil
 
+  override def dataSourceRewriteRules: Seq[Rule[LogicalPlan]] =
+    V2Writes :: Nil
```

Comment:
I would probably put this in the early pushdown batch, even though the name doesn't match. The rewrite batch needs to run before this so that writes created by it run `V2Writes` afterward. That's the same reason why early pushdown runs after plan rewrites.

Reply:
Ok, it makes sense after thinking more about it.